### ISSUE TITLE
Python 3.8+: NameError: name 'cache' is not defined

### ISSUE THREAD

**Issue by hugovk** [byexamples/byexample#116]
**Describe the bug**
A clear and concise description of what the bug is followed by the steps to reproduce it:
1. With Python 3.8, `pip install byexample` on macOS
2. Run `byexample -l python README.md`

```console
$ byexample -l python README.md
Process 0:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/byexample/common.py", line 191, in human_exceptions
    yield o
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/byexample/byexample.py", line 20, in execute_examples
    cache.synced(label=filename), \
NameError: name 'cache' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/byexample/jobs.py", line 24, in worker
    output.put(func(item, sigint_handler))
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/byexample/byexample.py", line 26, in execute_examples
    return executor.execute(examples, filename)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/byexample/common.py", line 201, in human_exceptions
    rlog.exception(msg=None, where=where_default)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/logging/__init__.py", line 1469, in exception
    self.error(msg, *args, exc_info=exc_info, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/logging/__init__.py", line 1463, in error
    self._log(ERROR, msg, args, **kwargs)
TypeError: _log() got an unexpected keyword argument 'where'
^C^C^C^C^C^C^C^C^C^C^C^C^C^C[1]    44820 terminated  byexample -l python README.md
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 6 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```
Needed to kill the Python process in Activity Monitor.

**Expected behavior**

Runs tests in the `README.md`

**byexample version**
Run ``byexample -V`` to know the version of ``byexample`` and ``python``.

```console
byexample -V
byexample 9.1.0 (Python 3.8.3) - GNU GPLv3

Write snippets of code in C++, Python, Ruby, and others as documentation and execute them as regression tests.

Copyright (C) Di Paola Martin - https://byexamples.github.io

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
```

I enabled Python 3.8 and 3.9-dev on the CI in my fork, and they failed (can ignore 3.10-dev, it fails due to dependencies not being ready yet):

https://travis-ci.com/github/hugovk/byexample/builds/171355524

**Comment by eldipa** [comment]
Hi, thank you for the report. It seems that there are two issues, close but unrelated:
 - `cache` is not defined
 - some tests fail in Python 3.8 and 3.9

For the latter, the problem is that  `byexample` uses a dark and undocumented feature of Python that is not longer available.
The feature is disabled by default and only enabled for testing so I disabled it for Python 3.8 and superior (release [9.1.1](https://github.com/byexamples/byexample/releases/tag/9.1.1)

The former issue is more weird. It says that `cache` is not defined. The variable is declared as global and the `main` function defines it *before* calling `execute_examples` where it fails.

https://github.com/byexamples/byexample/blob/9.1.0/byexample/byexample.py

Could you try to run the following snippet?
```python
import contextlib                                                                                                              
class Cache:                                                                                                                   
    @contextlib.contextmanager                                                                                                 
    def activated(self, val):                                                                                                  
        print("activating", val)                                                                                               
        try:                                                                                                                   
            yield self                                                                                                         
        finally:                                                                                                               
            print("deactivating", val)                                                                                         
                                                                                                                               
def bar():                                                                                                                     
    global cache                                                                                                               
    with cache.activated('bar'):                                                                                               
        print("I'm bar")                                                                                                       
                                                                                                                               
def foo():                                                                                                                     
    global cache                                                                                                               
    cache = Cache()                                                                                                            
    with cache.activated('foo'):                                                                                               
        print("I'm foo")                                                                                                       
        bar()                                                                                                                  
                                                                                                                               
if __name__ == '__main__':                                                                                                     
    foo()
```
Basically it does the same that `byexample` but with much less code. I'm suspecting an error in the Python interpreter that you have installed but we need a little evidence (I tested `byexample 9.1.1` in Python 3.8 it works so that's why I'm suspecting this) 

And thank you for your time filling the bug report!

**Comment by hugovk** [comment]
With that snippet I get this with Python 3.8.3:
```
activating foo
I'm foo
activating bar
I'm bar
deactivating bar
deactivating foo
```

**Comment by eldipa** [comment]
And that's the correct output....

It took me a while but I found at least the source of the problem.

`byexample` runs the tests in a separated worker using `multiprocessing`.
It executes `execute_examples` in a subprocess spawned from [here](https://github.com/byexamples/byexample/blob/9.1.1/byexample/byexample.py#L60).

`multiprocessing` has three methods to spawn a process: `spawn`, `fork` and `forkserver` and the default for macOS changed from `fork` to `spawn` in Python 3.8. [changes in multiprocessing](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods)

I managed to run `byexample` in macOS using `travis`. I copy the config for future references:
```
os: osx                                                                                                                        
language: sh                                                                                                                   
addons:                                                                                                                        
homebrew:                                                                                                                      
  update: true                                                                                                                 
  packages: python3                                                                                                            
before_install:                                                                                                                
  - pip3 install virtualenv                                                                                                    
  - virtualenv -p python3 ~/venv                                                                                               
  - source ~/venv/bin/activate
```
Unfortunately the installed version of python is 3.7 so to validate cause of the error I put this line just at the begin of the `main` function (with `multiprocessing` imported before):
```
multiprocessing.set_start_method('spawn')
```
And voila! `NameError: name 'cache' is not defined` https://travis-ci.org/github/byexamples/byexample/builds/698626680

Changing `spawn` by `fork` "solves" the problem but [open others](https://bugs.python.org/issue33725) of course.

And switching to `spawn` or even `forkserver` is not trivial. When `fork` mode is used, a subprocess is a exact copy of the parent process so objects like `cache`, `harvester` are inherited and available as copies.
Using another start mode will require to change that and pass the objects explicitly with a Queue probably; my concern however is with the console and logging.

@hugovk could you confirm me if the following steps fix the issue?:
 1. Import `multiprocessing` in `/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/byexample/byexample.py`
 1. Add `multiprocessing.set_start_method('fork')` at the begin of the `main` function, after the `global` statement in the same file than above.

**Comment by hugovk** [comment]
Aha, good investigation!

That reminds me, Flake8 recently disabled using multiprocessing recently for macOS/3.8+ due to the change:

* https://flake8.pycqa.org/en/latest/release-notes/3.8.0.html#a1-2020-04-24
* https://gitlab.com/pycqa/flake8/-/merge_requests/366



> @hugovk could you confirm me if the following steps fix the issue?:
> 
> 1. Import `multiprocessing` in `/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/byexample/byexample.py`
> 2. Add `multiprocessing.set_start_method('fork')` at the begin of the `main` function, after the `global` statement in the same file than above.

It now runs! Here's the output:

```console
$ byexample -l python README.md

**********************************************************************
File "README.md", line 86
Failed example:
    humanize.naturaldelta(delta, minimum_unit="milliseconds")
Expected:
'4 milliseconds'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: naturaldelta() got an unexpected keyword argument 'minimum_unit'


**********************************************************************
File "README.md", line 88
Failed example:
    humanize.naturaldelta(delta, minimum_unit="microseconds")
Expected:
'4000 microseconds'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: naturaldelta() got an unexpected keyword argument 'minimum_unit'


**********************************************************************
File "README.md", line 94
Failed example:
    humanize.naturaltime(delta, minimum_unit="milliseconds")
Expected:
'4 milliseconds ago'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: naturaltime() got an unexpected keyword argument 'minimum_unit'


**********************************************************************
File "README.md", line 96
Failed example:
    humanize.naturaltime(delta, minimum_unit="microseconds")
Expected:
'4000 microseconds ago'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: naturaltime() got an unexpected keyword argument 'minimum_unit'


**********************************************************************
File "README.md", line 116
Failed example:
    humanize.fractional(1/3)
Expected:
'1/3'
Got:
'0/1'


**********************************************************************
File "README.md", line 122
Failed example:
    humanize.fractional(0.333)
Expected:
'1/3'
Got:
'333/1000'


**********************************************************************
File "README.md", line 132
Failed example:
    humanize.scientific(0.3)
Expected:
'3.00 x 10⁻¹'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 134
Failed example:
    humanize.scientific(500)
Expected:
'5.00 x 10²'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 136
Failed example:
    humanize.scientific("20000")
Expected:
'2.00 x 10⁴'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 138
Failed example:
    humanize.scientific(1**10)
Expected:
'1.00 x 10⁰'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 140
Failed example:
    humanize.scientific(1**10, precision=1)
Expected:
'1.0 x 10⁰'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 142
Failed example:
    humanize.scientific(1**10, precision=0)
Expected:
'1 x 10⁰'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 153
Failed example:
    humanize.naturaltime(dt.timedelta(seconds=3))
Expected:
3 seconds ago
Got:
'3 seconds ago'


**********************************************************************
File "README.md", line 156
Failed example:
    humanize.naturaltime(dt.timedelta(seconds=3))
Expected:
3 секунды назад
Got:
'3 \xd1\x81\xd0\xb5\xd0\xba\xd1\x83\xd0\xbd\xd0\xb4\xd1\x8b \xd0\xbd\xd0\xb0\xd0\xb7\xd0\xb0\xd0\xb4'


**********************************************************************
File "README.md", line 159
Failed example:
    humanize.naturaltime(dt.timedelta(seconds=3))
Expected:
3 seconds ago
Got:
'3 seconds ago'


**********************************************************************
File "README.md", line 168
Failed example:
    humanize.i18n.activate("pt_BR")
Expected:
IOError: [Errno 2] No translation file found for domain: 'humanize'
Got:
<gettext.GNUTranslations instance at 0x105c56af0>


**********************************************************************
File "README.md", line 170
Failed example:
    humanize.i18n.activate("pt_BR", path="path/to/my/portuguese/translation/")
Expected:
<gettext.GNUTranslations instance ...>
Got:
<gettext.GNUTranslations instance at 0x105c56af0>


File README.md, 52/52 test ran in 0.20 seconds
[FAIL] Pass: 35 Fail: 17 Skip: 0
```

**Comment by eldipa** [comment]
Perfect, it is similar to the output that I've got.

Some are "innocent" errors:
 -  like `3 seconds ago` (the quotes are missing) and they're easy to fix
 - the one that fails with `IOError: [Errno 2]` it is tricky. You are saying that it should raise an exception but actually it is not doing that because `"pt_BR"` is a valid translation. You could set a non-existent one like `"xx_YY"` to test that `humanize` raises the expected exception or you could add `# byexample: +pass` to the example to say ["I don't care, assume that it is okay"](https://byexamples.github.io/byexample/basic/skip-and-pass).
 - the last example expects `<gettext.GNUTranslations instance ...>` but you could replace it with `<gettext.GNUTranslations <...>>`. The `<...>` will ["eat"](https://byexamples.github.io/byexample/basic/capture-and-paste) the `"instance at 0x105c56af0"` part which it is non-deterministic (actually I've got `"object at"` instead of `"instance at"`)

The rest are more serious (I didn't get those):
 - `TypeError` and `AttributeError` look like you are using a older version of `humanize`. `byexample` uses the same thing that you have installed. In fact, if you open a Python console you will see the same results that `byexample` does.
 - `'0/1'`, using `master` with `Python 3.5`, the test works just fine.
 - `'3 \xd1\x81\xd0....`, this is interesting, or there is a bug *or* `byexample` is using an incorrect encoding to interpret the output of `humanize` with [--encoding](https://byexamples.github.io/byexample/advanced/unicode)

**Comment by eldipa** [comment]
For future references about the `spawn/fork` issue: https://gitlab.com/pycqa/flake8/-/issues/587. It is exactly the same problem that `byexample` has.

**Comment by hugovk** [comment]
Yep, I was on an old version, I'd been bisecting. Here's the output with latest the master version, and on the branch with some README fixes (https://github.com/jmoiron/humanize/pull/134):


```console
$ python -c "import humanize; print(humanize.__version__)"
2.4.1.dev7
$ byexample -V
byexample 9.1.1 (Python 3.8.3) - GNU GPLv3
...
$ byexample -l python README.md

**********************************************************************
File "README.md", line 86
Failed example:
    humanize.naturaldelta(delta, minimum_unit="milliseconds")
Expected:
'4 milliseconds'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: naturaldelta() got an unexpected keyword argument 'minimum_unit'


**********************************************************************
File "README.md", line 88
Failed example:
    humanize.naturaldelta(delta, minimum_unit="microseconds")
Expected:
'4000 microseconds'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: naturaldelta() got an unexpected keyword argument 'minimum_unit'


**********************************************************************
File "README.md", line 94
Failed example:
    humanize.naturaltime(delta, minimum_unit="milliseconds")
Expected:
'4 milliseconds ago'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: naturaltime() got an unexpected keyword argument 'minimum_unit'


**********************************************************************
File "README.md", line 96
Failed example:
    humanize.naturaltime(delta, minimum_unit="microseconds")
Expected:
'4000 microseconds ago'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: naturaltime() got an unexpected keyword argument 'minimum_unit'


**********************************************************************
File "README.md", line 116
Failed example:
    humanize.fractional(1/3)
Expected:
'1/3'
Got:
'0/1'


**********************************************************************
File "README.md", line 132
Failed example:
    humanize.scientific(0.3)
Expected:
'3.00 x 10⁻¹'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 134
Failed example:
    humanize.scientific(500)
Expected:
'5.00 x 10²'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 136
Failed example:
    humanize.scientific("20000")
Expected:
'2.00 x 10⁴'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 138
Failed example:
    humanize.scientific(1**10)
Expected:
'1.00 x 10⁰'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 140
Failed example:
    humanize.scientific(1**10, precision=1)
Expected:
'1.0 x 10⁰'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 142
Failed example:
    humanize.scientific(1**10, precision=0)
Expected:
'1 x 10⁰'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'scientific'


**********************************************************************
File "README.md", line 156
Failed example:
    humanize.naturaltime(dt.timedelta(seconds=3))
Expected:
'3 секунды назад'
Got:
'3 \xd1\x81\xd0\xb5\xd0\xba\xd1\x83\xd0\xbd\xd0\xb4\xd1\x8b \xd0\xbd\xd0\xb0\xd0\xb7\xd0\xb0\xd0\xb4'


**********************************************************************
File "README.md", line 168
Failed example:
    humanize.i18n.activate("xx_XX")
Expected:
IOError: [Errno 2] No translation file found for domain: 'humanize'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/humanize/i18n.py", line 27, in activate
    translation = gettext_module.translation('humanize', path, [locale])
  File "/usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Versions/2.7/lib/python2.7/gettext.py", line 545, in translation
    raise IOError(ENOENT, 'No translation file found for domain', domain)
IOError: [Errno 2] No translation file found for domain: 'humanize'


**********************************************************************
File "README.md", line 170
Failed example:
    humanize.i18n.activate("xx_XX", path="path/to/my/own/translation/")
Expected:
<gettext.GNUTranslations <...>>
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/site-packages/humanize/i18n.py", line 27, in activate
    translation = gettext_module.translation('humanize', path, [locale])
  File "/usr/local/Cellar/python@2/2.7.17/Frameworks/Python.framework/Versions/2.7/lib/python2.7/gettext.py", line 545, in translation
    raise IOError(ENOENT, 'No translation file found for domain', domain)
IOError: [Errno 2] No translation file found for domain: 'humanize'


File README.md, 52/52 test ran in 0.21 seconds
[FAIL] Pass: 38 Fail: 14 Skip: 0
```

Something strange going on here with Python 2.7, so I uninstalled an old humanize from Python 2.7:

```console
$ python2 -m pip uninstall -y humanize
DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Found existing installation: humanize 0.5.1
Uninstalling humanize-0.5.1:
  Successfully uninstalled humanize-0.5.1
```
And it can't find the module at all:

```console
$ python -c "import humanize; print(humanize.__version__)"
2.4.1.dev7
$ byexample -l python README.md

**********************************************************************
File "README.md", line 38
Failed example:
    import humanize
Expected nothing
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named humanize


**********************************************************************
File "README.md", line 39
Failed example:
    humanize.intcomma(12345)
Expected:
'12,345'
Got:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'humanize' is not defined


**********************************************************************
...
File README.md, 52/52 test ran in 0.18 seconds
[FAIL] Pass: 4 Fail: 48 Skip: 0
```

So some confusion over Python interpreters. Perhaps byexample is using the Mac system version instead of the one in the terminal?

**Comment by eldipa** [comment]
That's correct. `byexample` uses `env` to run `python`.
In particular:
https://github.com/byexamples/byexample/blob/81050a09e41dc5f2a2a961dd65fcec18d973121c/byexample/modules/python.py#L476-L483
Which translates to:
`/usr/bin/env python -i`
You can tweak this with `--x-shebang`. For example, to get rid off the `env` part you could run:
```shell
$ byexample -x-shebang 'python:%p %a' -l python README.md
```
Or:
```shell
$ byexample -x-shebang 'python:python -i' -l python README.md
```
For python2, as you may guessed:
```shell
$ byexample -x-shebang 'python:python2 %a' -l python README.md
```

If you run `byexample -xh` it will print the help of the "advanced" options, including `--x-shebang`. [Docs here](https://byexamples.github.io/byexample/advanced/shebang)

To check that `byexample` is using the interpreter that you want, you can add an example to print it:

```python
>>> import sys, humanize
>>> print(sys.version)
>>> print(humanize.__version__)
```

(you need to put that snippet inside a markdown block in the README.md, like the rest of the examples. That's how `byexample` knows that it is python code and not a "quote", denoted by `>` in markdown.

In my case it prints:
```
<...>
3.5.3 (default, Sep 27 2018, 17:25:39)
[GCC 6.3.0 20170516]
<...>
2.4.1.dev5
```

I'm using a virtualenv for Python3 and installed `humanize` doing `pip install -e .` in the repository. Probably that's why it is using the expected `humanize` version.

Anyways, it is simpler to "override" the interpreter with `--x-shebang`.

**Comment by hugovk** [comment]
My `python` is aliased to `python3`, which is at `/Library/Frameworks/Python.framework/Versions/3.8/bin/python3`

How would I run it with that one?

**Comment by eldipa** [comment]
You can use `-x-shebang`:

```shell
$ byexample -x-shebang 
'python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 
%a' -l python README.md
```

Basically with `-x-shebang` you can override the binary and the 
arguments. The only thing required is that the binary needs to be 
compatible. `python3` is obviously compatible.

You can save the options in a file so you don't need to type them each 
time:

```shell
$ cat params
-l=python
-x-shebang=python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 
%a

$ byexample @params README.md
```

**Comment by eldipa** [comment]
You can use `-x-shebang`:

```shell
$ byexample -x-shebang 
'python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 %a' -l python README.md
```

Basically with `-x-shebang` you can override the binary and the 
arguments. The only thing required is that the binary needs to be 
compatible. `python3` is obviously compatible.

You can save the options in a file so you don't need to type them each 
time:

```shell
$ cat params
-l=python
-x-shebang=python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 %a

$ byexample @params README.md
```

**Comment by hugovk** [comment]
Thanks, now I get this:

```console
$ byexample -x-shebang 'python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 %a' -l python README.md

**********************************************************************
File "README.md", line 38
Failed example:
    import humanize
Expected nothing
Got:
import humanize


**********************************************************************
File "README.md", line 39
Failed example:
    humanize.intcomma(12345)
Expected:
'12,345'
Got:
humanize.intcomma(12345)
'12,345'


**********************************************************************
File "README.md", line 41
Failed example:
    humanize.intword(123455913)
Expected:
'123.5 million'
Got:
humanize.intword(123455913)
'123.5 million'


**********************************************************************
...
File README.md, 55/55 test ran in 0.72 seconds
[FAIL] Pass: 0 Fail: 55 Skip: 0
```

![image](https://user-images.githubusercontent.com/1324225/85109816-9bbd7500-b21a-11ea-8977-a27d237b2626.png)

**Comment by eldipa** [comment]
Okay, one step forward, one step back.

When you type something in a terminal you can see what you typed because 
the terminal echo it back to you.

`byexample` "forces" to not do it that so it can read only the output 
coming from the example that it is executing.

But it seems that 
`/Library/Frameworks/Python.framework/Versions/3.8/bin/python3` (I will 
call it `fpython3` for short) is ignoring the settings.

I found this problem with the interpreter for C/C++ (`cling`) so I have 
a workaround but I'm going to need a few days to do the refactor and 
make it visible to the user.

My knowledge of macos is basically zero. Is this `fpython3` something 
standard? I wonder why is happening because I ran the tests in osx with 
Travis and it worked well there. Is any special or different from 
a standard setup?

**Comment by hugovk** [comment]
It's the Python from https://www.python.org/downloads/:

* https://www.python.org/ftp/python/3.8.3/python-3.8.3-macosx10.9.pkg

I also have older Pythons installed via pyenv, but the python.org installer will be more common.

**Comment by eldipa** [comment]
It is interesting that `/Library/Frameworks/Python.framework/Versions/3.8/bin/python3` is ignoring the `echo = false`. May be is how the terminal is interpreting this but what it is confusing me is that if you use `python` directly that doesn't happen (and the terminal is the same, you just changed the interpreter).

Anyways, I released `9.2.0` which brings the `+force-echo-filtering` option. The details are[here](https://byexamples.github.io/byexample/advanced/echo-filtering).

Basically you should run `byexample` like this (long version):

```
$ byexample -x-shebang 'python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 %a' -l python -o '+force-echo-filtering' README.md
```

Short version would be:

```
$ cat params
-l=python
-x-shebang=python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 %a
-o=+force-echo-filtering

$ byexample @params README.md
```

If you have time, I would like to know if that works in your environment.

**Comment by hugovk** [comment]
Upgrading to 9.2.0, and re-applying `multiprocessing.set_start_method('fork')` from https://github.com/byexamples/byexample/issues/116#issuecomment-644275839, the long version gives this:

```console
$ byexample -x-shebang 'python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 %a' -l python -o '+force-echo-filtering' README.md

**********************************************************************
File "README.md", line 66
Failed example:
    humanize.naturaltime(dt.datetime.now() - dt.timedelta(seconds=1))
Expected:
'a second ago'
Got:
=1))
'a second ago'


**********************************************************************
File "README.md", line 68
Failed example:
    humanize.naturaltime(dt.datetime.now() - dt.timedelta(seconds=3600))
Expected:
'an hour ago'
Got:
=3600))
'an hour ago'


**********************************************************************
File "README.md", line 168
Failed example:
    humanize.i18n.activate("pt_BR")
Expected:
IOError: [Errno 2] No translation file found for domain: 'humanize'
Got:
<gettext.GNUTranslations object at 0x10b7f6e20>


**********************************************************************
File "README.md", line 170
Failed example:
    humanize.i18n.activate("pt_BR", path="path/to/my/portuguese/translation/")
Expected:
<gettext.GNUTranslations instance ...>
Got:
ranslation/")
<gettext.GNUTranslations object at 0x10b7f6e20>


File README.md, 52/52 test ran in 0.82 seconds
[FAIL] Pass: 48 Fail: 4 Skip: 0
```

**Comment by eldipa** [comment]
Perfect, as you see there are some "spurious" strings like `=1))`and `=3600))`. Those are because `+force-echo-filtering` uses a terminal emulator and this also put some limits to the terminal.

Basically, you are seeing the output through a "small" terminal. The solution is to expand the terminal with [+geometry](https://byexamples.github.io/byexample/advanced/geometry):

```shell
$ cat params
-l=python
-x-shebang=python:/Library/Frameworks/Python.framework/Versions/3.8/bin/python3 %a
-o=+force-echo-filtering +geometry=24x300

$ byexample @params README.md
```

In this case, your emulated terminal will have 24 rows, 300 columns. That should be more than enough for the output of your examples.

**Comment by hugovk** [comment]
All good now, thank you!

**Comment by eldipa** [comment]
No, thank *you*. I think nobody used so many features of `byexamples` like you. It was a pleasure.
If you are okay, I will close this ticket.

**Comment by hugovk** [comment]
You're welcome!
